mobile_verifier/src/verifier.rs

```diff
@@ -1,148 +1,223 @@
-    pub speedtest_avgs: FileSinkClient,
-    pub radio_rewards: FileSinkClient,
+    pub heartbeats: Receiver<FileInfoStream<CellHeartbeatIngestReport>>,
+    pub speedtests: Receiver<FileInfoStream<CellSpeedtestIngestReport>>,
+    pub valid_heartbeats: FileSinkClient,
+    pub valid_speedtests: FileSinkClient,
     pub mobile_rewards: FileSinkClient,
     pub reward_manifests: FileSinkClient,
+    pub data_transfer_ingest: FileStore,
     pub reward_period_hours: i64,
-    pub verifications_per_period: i32,
-    pub verification_offset: Duration,
-    pub verifier: Verifier,
+    pub config_client: Client,
     pub price_tracker: PriceTracker,
-    pub data_transfer_ingest: FileStore,
 }
 
 impl VerifierDaemon {
-    pub async fn run(mut self, shutdown: &triggered::Listener) -> anyhow::Result<()> {
+    pub async fn run(self, shutdown: &triggered::Listener) -> anyhow::Result<()> {
         tracing::info!("Starting verifier service");
 
-        let reward_period_length = Duration::hours(self.reward_period_hours);
-        let verification_period_length = reward_period_length / self.verifications_per_period;
+        let Self {
+            pool,
+            mut heartbeats,
+            mut speedtests,
+            valid_heartbeats,
+            valid_speedtests,
+            config_client,
+            mobile_rewards,
+            reward_manifests,
+            price_tracker,
+            data_transfer_ingest,
+            reward_period_hours,
+        } = self;
+
+        let heartbeat_pool = pool.clone();
+        let heartbeat_config_client = config_client.clone();
+        let mut heartbeat_verification_task: JoinHandle<anyhow::Result<()>> =
+            tokio::spawn(async move {
+                let heartbeat_cache = Arc::new(Cache::<(String, DateTime<Utc>), ()>::new());
+
+                let cache_clone = heartbeat_cache.clone();
+                tokio::spawn(async move {
+                    cache_clone
+                        .monitor(4, 0.25, std::time::Duration::from_secs(60 * 60 * 3))
+                        .await
+                });
+
+                loop {
+                    let Some(file) = heartbeats.recv().await else {
+                        bail!("Heartbeat report file stream was dropped unexpectedly");
+                    };
+                    let epoch =
+                        (file.file_info.timestamp - Duration::hours(1))..file.file_info.timestamp;
+                    let mut transaction = heartbeat_pool.begin().await?;
+                    let reports = file.into_stream(&mut transaction).await?;
+
+                    let mut validated_heartbeats = pin!(
+                        Heartbeat::validate_heartbeats(&heartbeat_config_client, reports, &epoch)
+                            .await
+                    );
+
+                    while let Some(heartbeat) = validated_heartbeats.next().await.transpose()? {
+                        heartbeat.write(&valid_heartbeats).await?;
+                        /*
```

Comment:
should these commented out sections relating to cache usage be removed ?

Reply:
This cache still feels very valuable as it will stop us from hitting the database for a large number of heartbeats.